### PR TITLE
Add IRS_ASSERT

### PR DESCRIPTION
**Pull request by MBkkt**

### CODE REVIEW

File: core/utils/type_utils.hpp [hunk 1]
```diff
@@ -37,7 +37,7 @@ namespace iresearch {
 ////////////////////////////////////////////////////////////////////////////////
 template<typename T>
 constexpr std::string_view ctti() noexcept {
-  return {IRESEARCH_CURRENT_FUNCTION};
+  return {IRS_FUNC_NAME};
```
Comment (MBkkt) [inline core/utils/type_utils.hpp:40]: .

File: core/utils/numeric_utils.cpp [hunk 2]
```diff
@@ -142,7 +142,7 @@ typename EncodeTraits::type decode(const byte_type* in) {
   return value;
 }
 
-// assert that signed right shift works as expected
+// IRS_ASSERT that signed right shift works as expected
```
Comment (MBkkt) [inline core/utils/numeric_utils.cpp:145]: .

File: core/utils/assert.hpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+#pragma once
+
+#include <cstddef>
+#include <string_view>
+
+#include "shared.hpp"
+
+namespace iresearch {
+
+enum class LogLevel : unsigned char {
```
Comment (Dronplane) [inline core/utils/assert.hpp:32]: Why not use log levels/log callback from log.hpp/cpp ? We could just add here additional level for assertion if needed (or just use IRL_FATAL)

Comment (MBkkt) [reply core/utils/assert.hpp:32]: I don't want add new global value

Comment (Dronplane) [reply core/utils/assert.hpp:32]: Then why  enum? Are we expecting here many asserion log levels?

File: core/utils/assert.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  if (const auto callback = sCallbacks[static_cast<std::size_t>(level)];
+      IRS_LIKELY(callback != nullptr)) {
+    callback(file, line, func, condition);
+  }
+}
+
+}  // namespace detail
+
+void SetCallback(LogLevel level, LogCallback callback) noexcept {
+  detail::sCallbacks[static_cast<std::size_t>(level)] = callback;
```
Comment (Dronplane) [inline core/utils/assert.cpp:45]: Suggest validating `level` value here. As this value comes from outside of the library. If we decide in the end to have multiple assertion levels

Comment (MBkkt) [reply core/utils/assert.cpp:45]: It's enum class, why it should be validated?

**Comment by gnusi** [comment]
Please set callback in `iresearch/utils/main.cpp`  as well

File: tests/tests_main.cpp [hunk 5]
```diff
@@ -362,11 +362,15 @@ void test_base::SetUp() {
...
 }
 
-// -----------------------------------------------------------------------------
-// --SECTION--                                                              main
-// -----------------------------------------------------------------------------
+static void AssertCallback(std::string_view file, std::size_t line,
+                           std::string_view function,
+                           std::string_view condition) noexcept {
+  FAIL() << file << ":" << line << ": " << function << ": Condition '"
+         << condition << "' is true.";
```
Comment (gnusi) [inline tests/tests_main.cpp:369]: maybe `is false`?

File: core/formats/formats_10.cpp [hunk 6]
```diff
@@ -2904,8 +2892,9 @@ void segment_meta_writer::write(directory& dir, std::string& meta_file,
 }
 
 struct segment_meta_reader final : public irs::segment_meta_reader {
-  virtual void read(const directory& dir, segment_meta& meta,
-                    std::string_view filename = {}) override;  // null == use meta
+  virtual void read(
```
Comment (gnusi) [inline core/formats/formats_10.cpp:2895]: remove virtual

File: core/formats/formats_10.cpp [hunk 7]
```diff
@@ -2629,9 +2616,10 @@ struct index_meta_reader final : public irs::index_meta_reader {
   virtual bool last_segments_file(const directory& dir,
                                   std::string& name) const override;
 
-  virtual void read(const directory& dir, index_meta& meta,
-                    std::string_view filename = {}) override;  // null == use meta
-};                                                       // index_meta_reader
+  virtual void read(
```
Comment (gnusi) [inline core/formats/formats_10.cpp:2619]: remove virtual

File: core/shared.hpp [hunk 8]
```diff
@@ -102,118 +60,41 @@
...
-#define GCC_ONLY(...) __VA_ARGS__
-#else
-#define GCC_ONLY(...)
-#endif
-
 // check if sizeof(float_t) == sizeof(double_t)
 #if defined(FLT_EVAL_METHOD) && \
   ((FLT_EVAL_METHOD == 1) || (FLT_EVAL_METHOD == 2))
 static_assert(sizeof(float_t) == sizeof(double_t),
               "sizeof(float_t) != sizeof(double_t)");
```
Comment (gnusi) [inline core/shared.hpp:67]: also remove message?

File: core/shared.hpp [hunk 9]
```diff
@@ -230,22 +111,6 @@ static_assert(sizeof(float_t) != sizeof(double_t),
 #define IRESEARCH_SSE2
 #endif
 
-#ifdef __SSE4_1__
```
Comment (gnusi) [inline core/shared.hpp:233]: I prefer to keep it for future

File: core/shared.hpp [hunk 10]
```diff
@@ -102,118 +60,41 @@
...
-#ifndef __has_attribute
-#define IRESEARCH_HAS_ATTRIBUTE(x) 0
-#else
-#define IRESEARCH_HAS_ATTRIBUTE(x) __has_attribute(x)
-#endif
-
-// IRESEARCH_ATTRIBUTE_NONNULL
-#if IRESEARCH_HAS_ATTRIBUTE(nonnull) || \
-  (defined(__GNUC__) && !defined(__clang__))
+#if __has_attribute(nonnull)
```
Comment (gnusi) [inline core/shared.hpp:97]: what if __has_attribute isn't defined?

File: core/shared.hpp [hunk 11]
```diff
@@ -102,118 +60,41 @@
...
-#define IRESEARCH_CURRENT_FUNCTION __PRETTY_FUNCTION__
+#define IRS_FUNC_NAME __FUNCSIG__
 #else
-#error "compiler is not supported"
+#define IRS_FUNC_NAME __func__
 #endif
-
-// IRESEARCH_COMPILER_HAS_FEATURE
-#ifndef __has_feature
-#define IRESEARCH_COMPILER_HAS_FEATURE(x) \
```
Comment (gnusi) [inline core/shared.hpp:201]: let's keep it

File: core/shared.hpp [hunk 12]
```diff
@@ -102,118 +60,41 @@
...
-#define IRESEARCH_PLUGIN_EXPORT extern "C" IRESEARCH_HELPER_DLL_EXPORT
-#else
-#define IRESEARCH_PLUGIN_EXPORT
-#endif  // IRESEARCH_DLL_PLUGIN
-#else   // IRESEARCH_DLL is not defined: this means IRESEARCH is a static lib.
-#define IRESEARCH_API
-#endif  // IRESEARCH_DLL
-
 // MSVC 2015 does not define __cpp_lib_generic_associative_lookup macro
 #if (defined(__cpp_lib_generic_associative_lookup) || \
```
Comment (gnusi) [inline core/shared.hpp:82]: I don't think we need it

File: core/shared.hpp (hunk 10 above)
Comment (MBkkt) [reply core/shared.hpp:97]: It always defined in supported compilers

File: core/shared.hpp (hunk 11 above)
Comment (MBkkt) [reply core/shared.hpp:201]: Why?

File: core/shared.hpp (hunk 12 above)
Comment (MBkkt) [reply core/shared.hpp:82]: Ok

File: core/utils/assert.hpp (hunk 3 above)
Comment (MBkkt) [reply core/utils/assert.hpp:32]: I think maybe in future we remove log with printf format style

File: tests/tests_main.cpp (hunk 5 above)
Comment (MBkkt) [reply tests/tests_main.cpp]: No, true, because LOG_MESSAGE(condition) condition should be true
assert add `!`

File: core/shared.hpp (hunk 11 above)
Comment (MBkkt) [reply core/shared.hpp:201]: Anyway done

File: core/shared.hpp (hunk 10 above)
Comment (MBkkt) [reply core/shared.hpp]: I don't think it's necessary, but ok

File: core/utils/assert.hpp (hunk 3 above)
Comment (MBkkt) [reply core/utils/assert.hpp:32]:
```cpp
#define LOG_ERROR(cond, message)
 do {
   if (cond) {
     LogMessage(source_location, #cond, message)
   }
 } while(false)


// message will be computed only if condition true
LOG_ERROR(result.isError(), absl::StrCat/Substitute(...)); 
```

Comment (Dronplane) [reply core/utils/assert.hpp:32]: Still don't get it, why we need separate log levels enum for assertions, with just single value in it. If we will replace printf style logger (I'm totally for it) we will add proper log levele there.

File: tests/tests_main.cpp (hunk 5 above)
Comment (Dronplane) [reply tests/tests_main.cpp]: Then we could supply condition as string as separate macros arg and avoid this  "!<condition> is true"
Like this:
#define ASSERT_PRINTER(cond, cond_text)                           \
  do {                                                           \
    if (!(cond)) {                                              \
      std::cout <<  __FILE__ << ", " << __LINE__ << "[" << cond_text << "] is false" << std::endl;     \
    }                                                            \
  } while (false)

#define IRS_ASSERT(cond) ASSERT_PRINTER(cond, #cond)

Comment (MBkkt) [reply tests/tests_main.cpp]: I don't think formatting in macro is good idea, anyway fixed

File: core/utils/assert.hpp [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+  do {                                                                      \
+    if (!(cond)) {                                                          \
+      ::iresearch::detail::AssertMessage(__FILE__, __LINE__, IRS_FUNC_NAME, \
+                                         #cond, message);                   \
+    }                                                                       \
+  } while (false)
+
+#define IRS_GET_MACRO(arg1, arg2, MACRO, ...) MACRO
+
+#ifdef NDEBUG
```
Comment (Dronplane) [inline core/utils/assert.hpp:61]: Should this also be controlled by `IRESEARCH_DEBUG` ?

Comment (MBkkt) [reply core/utils/assert.hpp:61]: No

Comment (MBkkt) [reply core/utils/assert.hpp:61]: This just remove unnecessary if (false) etc in release

Comment (Dronplane) [reply core/utils/assert.hpp:61]: But RelWithDebugInfo also defines `NDEBUG`.
I was expecting that whole assertions should either trigger or not based only on `IRESEARCH_DEBUG`

Comment (MBkkt) [reply core/utils/assert.hpp]: >But RelWithDebugInfo also defines NDEBUG.

And it's good

Comment (MBkkt) [reply core/utils/assert.hpp]: > I was expecting that whole assertions should either trigger or not based only on IRESEARCH_DEBUG

Yes, NDEBUG defines only IRS_STUB implementation

**Review by Dronplane — APPROVED** [review]
LGTM

**Pull request merged**